sql/src/main/java/io/crate/analyze/where/DocKeys.java

```diff
@@ -0,0 +1,170 @@
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.analyze.where;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.collect.Lists;
+import io.crate.planner.symbol.Literal;
+import io.crate.planner.symbol.Symbol;
+import io.crate.planner.symbol.ValueSymbolVisitor;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.Base64;
+import org.elasticsearch.common.Preconditions;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+
+public class DocKeys implements Iterable<DocKeys.DocKey> {
+
+    private final int width;
+    private int clusteredByIdx;
+    private final boolean withVersions;
+    private final List<List<Symbol>> docKeys;
+    private final List<Integer> partitionIdx;
+
+    public class DocKey {
+
+        private final List<Symbol> key;
+
+        public DocKey(int pos) {
+            key = docKeys.get(pos);
+        }
+
+        @Nullable
+        public Optional<Long> version() {
+            if (withVersions && key.get(width) != null) {
+                return Optional.of(((Literal<Long>) key.get(width)).value());
+            }
+            return Optional.absent();
+        }
+
+        public String routing() {
+            if (clusteredByIdx >= 0){
+                return ValueSymbolVisitor.STRING.process(key.get(clusteredByIdx));
+            } else {
+                return id();
+            }
+
+        }
+
+        public Optional<List<BytesRef>> partitionValues(){
+            if (partitionIdx == null || partitionIdx.isEmpty()) {
+                return Optional.absent();
+            }
+            List<BytesRef> values = Lists.transform(partitionIdx, new Function<Integer, BytesRef>() {
+                @Nullable
+                @Override
+                public BytesRef apply(@Nullable Integer input) {
+                    return ValueSymbolVisitor.BYTES_REF.process(key.get(input));
+                }
+            });
+            return Optional.of(values);
+        }
+
+        public String id() {
```

Comment:
Should we cache the result here? Calling id() multiple times might be a bit expensive otherwise.